реализация ТЗ14: Add controllers

cepheil opened the pull request.

cepheil committed:
feat(user): реализован модуль управления пользователями

- добавлены DTO: UserCreateDto, UserDto, UserUpdateDto
- реализован маппер UserMapper
- добавлен интерфейс UserRepository и InMemoryUserRepository
- реализован сервисный слой: UserService, UserServiceImpl
- создан контроллер UserController с CRUD-операциями

cepheil committed:
feat(item): реализован модуль управления вещами

- добавлены DTO: ItemCreateDto
                 ItemDto
                 ItemUpdateDto
- реализован маппер ItemMapper
- добавлен интерфейс ItemRepository и InMemoryItemRepository
- реализован сервисный слой: ItemService, ItemServiceImpl
- создан контроллер ItemController с CRUD-операциями

feat(exception):
- добавлены исключения: ConflictException
                        ForbiddenException
                        NotFoundException
                        ValidationException
- реализован глобальный ErrorHandler, ErrorResponse

avfyodorov commented on src/main/java/ru/practicum/shareit/exception/ErrorHandler.java:
```diff
@@ -0,0 +1,54 @@
+        log.warn("NotFoundException: {}", e.getMessage());
+        return buildErrorResponse(e);
+    }
+
```
Так же можно было бы  добавить обработчик для исключения, которое выбрасывается при валидации данных с помощью аннотаций - MethodArgumentNotValidException, возвращать будет статус 400 BAD_REQUEST, так как ошибка связана с некорректными данными от пользователя.

avfyodorov commented on src/main/java/ru/practicum/shareit/exception/ErrorHandler.java:
```diff
@@ -0,0 +1,54 @@
+        log.warn("ValidationException: {}", e.getMessage());
+        return buildErrorResponse(e);
+    }
+
```
Кроме  MethodArgumentNotValidException хорошо было бы добавить и ConstraintViolationException,  как раз эти исключения возникают при ошибках валидации.

avfyodorov commented on src/main/java/ru/practicum/shareit/item/mapper/ItemMapper.java:
```diff
@@ -0,0 +1,55 @@
+import ru.practicum.shareit.item.model.Item;
+
+
+public class ItemMapper {
```
В принципе, для мапперов можно добавить аннотацию: 

`@NoArgsConstructor(access = AccessLevel.PRIVATE) или @UtilityClass`

Поскольку все методы Статик, то нет никакого смысла разрешать  создавать экземпляры этого класса. 
 необязательно,  на Ваше усмотрение.

avfyodorov commented on src/main/java/ru/practicum/shareit/user/mapper/UserMapper.java:
```diff
@@ -0,0 +1,42 @@
+import ru.practicum.shareit.user.dto.UserDto;
+import ru.practicum.shareit.user.dto.UserUpdateDto;
+
+public class UserMapper {
```
можно добавить аннотацию: 

`@NoArgsConstructor(access = AccessLevel.PRIVATE) или @UtilityClass`

avfyodorov reviewed (commented):
Добрый вечер, Сергей! 

Очень хорошая работа, от меня всего пара уточнений.

cepheil committed:
refactor: доработан ErrorHandler и добавлены ограничения для мапперов

- обновлён ErrorHandler: поддержка MethodArgumentNotValidException и ConstraintViolationException

- мапперам добавлены @NoArgsConstructor(access = AccessLevel.PRIVATE), чтобы запретить создание экземпляров

avfyodorov reviewed (commented):
Добрый вечер, Сергей!

Замечаний нет.
Работа принята.
